[OD-1784] Handle NotFound and NotAuthorized exceptions

jguo144 opened the pull request:
This PR adds exception handling for NotFound and NotAuthorized errors.
The fix is to return a 400 or 404 status when users try to download a file that is no longer in CKAN or do not have permission to download.

For other events like viewing a dataset, resource, or organization the extension will catch all exception and pass so that the default controller or view can handle things.

This PR also fixes an issue where `new` is treated as an id instead of the route to create datasets or groups. To fix this we add url rules to point back to the default dataset view or group view.

jguo144 committed:
Handle NotFound and NotAuthorized exceptions

jguo144 committed:
Fix new org/dataset error and add more try catches

peterVorman commented on ckanext/googleanalytics/controller.py:
```diff
@@ -181,14 +181,14 @@ def _post_analytics(self, user, request_obj_type, request_function, request_id):
         # We do not want to perform read operation on organization id "new",
         # where it results in a NotFound error
-        if id != "new":
+        if id and id != "new":
```
we can reverse logic here. 
```
        if id and id == "new":
            return OrganizationController.new(self)
       try:
                org = toolkit.get_action('organization_show')({},{'id':id})
                org_title = org.get('title')
                self._post_analytics(c.user, "Organization", "View", org_title)
         except (toolkit.ObjectNotFound, toolkit.NotAuthorized):
                pass
        return OrganizationController.read(self, id, limit=20)
```

peterVorman commented on ckanext/googleanalytics/controller.py:
```diff
@@ -237,7 +237,7 @@ def read(self, id):
         # If we simply return PackageController.read() or return w/o a
         # PackageController.new() operation, a blank page or error page will appear
-        else:
+        elif id == "new":
```
the same reverse here.

jguo144 committed:
Refactor id conditional logic

Impicitly check that id equals id. Also catch more exceptions and pass to default controller/view.

peterVorman commented on ckanext/googleanalytics/controller.py:
```diff
@@ -181,15 +181,15 @@ def _post_analytics(self, user, request_obj_type, request_function, request_id):
+            self._post_analytics(c.user, "Organization", "View", org_title)
+        except Exception as e:
+            log.debug(e)
+            pass
```
'pass' is not necessary here.

peterVorman commented on ckanext/googleanalytics/controller.py:
```diff
@@ -231,29 +231,35 @@ def _post_analytics(self, user, request_obj_type, request_function, request_id):
-            return PackageController.new(self)
+        except Exception as e:
+            log.debug(e)
+            pass
```
'pass' is not necessary here.

peterVorman commented on ckanext/googleanalytics/controller.py:
```diff
@@ -231,29 +231,35 @@ def _post_analytics(self, user, request_obj_type, request_function, request_id):
+                self._post_analytics(c.user, "Organization", "View", org_id)
+        except Exception as e:
+            log.debug(e)
+            pass
```
'pass' is not necessary here.

peterVorman commented on ckanext/googleanalytics/controller.py:
```diff
@@ -231,29 +231,35 @@ def _post_analytics(self, user, request_obj_type, request_function, request_id):
+            org_id = package.get('organization', {}).get('title')
+        except Exception as e:
+            log.debug(e)
+            pass
```
'pass' is not necessary here.

jguo144 committed:
Remove unnecessary lines

jguo144 committed:
Fix new resources page

peterVorman reviewed (approved):
LGTM
